### PR TITLE
WIP: Add apr

### PR DESCRIPTION
**Pull request by jakirkham**
Fixes https://github.com/conda-forge/staged-recipes/issues/1662

Add the Apache Portable Runtime as a `conda` package. Needed to build `svn`.

**Todo:**

* [x] Figure out what to do with `apr-util` and `apr-iconv`. (appears `defaults` bundled them 😕)
* [x] Determine the XML situation.
* [ ] Handle `apr-iconv` on Windows.

### CODE REVIEW

File: recipes/apr/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+source:
+  fn: {{ name }}-{{ version }}.tar.gz
+  url: http://mirror.metrocast.net/apache/apr/apr-{{ version }}.tar.gz
+  sha256: 1af06e1720a58851d90694a984af18355b65bb0d047be03ec7d659c746d6dbdb
+  patches:
+    # Skip a failing test on Mac due to `poll`'s behavior.
+    #
+    # ref: https://www.mail-archive.com/dev@apr.apache.org/msg26625.html
+    #
+    - macos_testpoll.c.patch
```
Comment (jakirkham) [inline recipes/apr/meta.yaml]: Needed this patch to workaround Mac's `poll` (mis)behavior, which causes test failures. Please see the ref below or just google. 😒

ref: https://www.mail-archive.com/dev@apr.apache.org/msg26625.html

File: recipes/apr/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+source:
+  fn: {{ name }}-{{ version }}.tar.gz
+  url: https://archive.apache.org/dist/{{ name }}/{{ name }}-{{ version }}.tar.bz2
+  sha256: 7d03ed29c22a7152be45b8e50431063736df9e1daa1ddf93f6a547ba7a28f67a
+  patches:
+    # Skip a failing test on Mac due to `poll`'s behavior.
+    #
+    # ref: https://www.mail-archive.com/dev@apr.apache.org/msg26625.html
+    #
+    - macos_testpoll.c.patch  # [osx]
```
Comment (jakirkham) [inline recipes/apr/meta.yaml]: Needed this patch to workaround Mac's `poll` (mis)behavior, which causes test failures. Please see the ref below or just google. 😒

ref: https://www.mail-archive.com/dev@apr.apache.org/msg26625.html

File: recipes/apr/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    # Apply a patch to fix a double free issue on Windows.
+    #
+    # ref: https://bz.apache.org/bugzilla/show_bug.cgi?id=59711
+    #
+    - win_apr_app.c.patch     # [win]
```
Comment (jakirkham) [inline recipes/apr/meta.yaml]: Needed this patch to fix a double free issue on Windows. This also caused a test failure. More details in the bug report linked below.

ref: https://bz.apache.org/bugzilla/show_bug.cgi?id=59711

**Comment by jakirkham** [comment]
So `defaults` appears to bundle `apr`, `apr-utils`, and `apr-iconv` together. Personally I'd rather see them as separate packages. This is the common way Linux distros and Homebrew handle this. Though I don't see many package managers providing `apr-iconv` probably because they just use `libiconv` instead. As we don't want to break things with `defaults`, I'd like to suggest that Apache's APR simply be `libapr` and `apr` be a metapackage that pulls in the other dependencies. Thoughts on this?

cc @mingwandroid @msarahan @ilanschnell

**Comment by mingwandroid** [comment]
Sounds like a fine plan to me.

**Comment by jakirkham** [comment]
So it seems like building `apr-iconv` on Windows is a bear as it doesn't use CMake. I don't think I'll have time to figure this out. If someone wants to push this forward great, but I plan on leaving this as an issue on the feedstock. Given we have `libiconv` on Windows and `apr-iconv` exists as a substitute for when `libiconv` is not available and `apr-iconv` has been untouched for a decade, I'm not sure how much this matters.

The rest of these build fine on Mac and Linux. Also `apr-util` should build fine on Windows, but will have to wait for the CIs to know for sure.

**Comment by jakirkham** [comment]
Also should add this quote...

> As announced previously, the 0.9.x branches of Apache Portable Runtime library, Apache Portable Runtime Utility library, and the companion APR-iconv library have been retired. No further bug or security fixes will be available for these branches. 

ref: http://www.apache.org/dist/apr/Announcement1.x.html

**Comment by scopatz** [comment]
@jakirkham is this still a WIP?

**Comment by jakirkham** [comment]
Yeah the Windows situation is kind of a mess. Unfortunately the Windows issues cannot be cleanly separated from the package structure. IOW skipping Windows will not fix the problem and will make it harder to fix later. We may need `conda-build` 3 to get this to work reasonably. Namely we need to be able to make multiple build artifacts from one recipe/build.

As this is pretty low priority for me in the near term, `defaults` has a copy, and it is difficult to unentangle currently, have opted to let it sit.

**Comment by scopatz** [comment]
OK No worries!  I was just going through the old PRs that have passed CI.

**Comment by jakirkham** [comment]
np. I don't think I clearly stated the blocker before. So it is good to have it articulated somewhere.

**Comment by jakirkham** [comment]
Replacing with PR ( https://github.com/conda-forge/staged-recipes/pull/5763 ).

**Pull request closed without merging**